### PR TITLE
build: start only the minimum number of Saucelabs browsers required

### PR DESCRIPTION
**Pull request by gregmagolan**
This should save on Saucelabs resources so that if only one saucelabs test is run then only one set of browsers will be started.

Tested locally that ,

- `./scripts/test/run-saucelabs-tests.sh 1 --nocache_test_results` starts just 1 browser
- `./scripts/test/run-saucelabs-tests.sh 8 --nocache_test_results` starts 8 browsers one at a time
- `./scripts/test/run-saucelabs-tests.sh 100 --nocache_test_results --jobs=4` starts just 4 browsers since that is the maximum # of parallel tests run

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -712,6 +712,9 @@ workflows:
       - test_angular_devtools:
           requires:
             - setup
+      - saucelabs:
```
Comment (gregmagolan) [inline .circleci/config.yml:715]: remove before landing

File: tools/saucelabs-daemon/background-service/saucelabs-daemon.ts [hunk 2]
```diff
@@ -294,16 +299,31 @@ export class SaucelabsDaemon {
...
+   * Given a browserId, returns a browser that matches the browserId and is free
+   * or launching with no pending test. If no such browser if found, returns
+   * undefined.
    **/
-  private _findMatchingBrowsers(browserId: string): RemoteBrowser[] {
-    const browsers: RemoteBrowser[] = [];
-    this._activeBrowsers.forEach(b => {
-      if (b.id === browserId) {
-        browsers.push(b);
+  private _findAvailableBrowser(browserId: string): RemoteBrowser|undefined {
```
Comment (devversion) [inline tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:306]:
```suggestion
  private _findAvailableBrowser(browserId: string): RemoteBrowser|null {
```

File: tools/saucelabs-daemon/background-service/saucelabs-daemon.ts [hunk 3]
```diff
@@ -69,11 +70,15 @@ export class SaucelabsDaemon {
...
 
+  /* Number of parallel executions started */
+  private _parallelExecutions: number = 0;
+
   constructor(
       private _username: string,
       private _accessKey: string,
       private _buildName: string,
-      private _browsers: Browser[],
+      private _customLaunchers: CustomLaunchers,
```
Comment (devversion) [inline tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:80]: Is `Browser[]` not sufficient here anymore?

File: tools/saucelabs-daemon/background-service/saucelabs-daemon.ts [hunk 4]
```diff
@@ -154,36 +159,30 @@ export class SaucelabsDaemon {
...
-        } else {
-          this._pendingTests.set(browser, test);
-          return true;
-        }
+      await this.launchBrowsers();
+
+      browser = this._findAvailableBrowser(test.requestedBrowserId);
+      if (!browser) {
+        // Still no browser means we are likely not requesting a browser type that is being
+        // launched. Logic error?
```
Comment (devversion) [inline tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:174]: maybe log something? and we should remove the `Logic error?` question I think

File: tools/saucelabs-daemon/background-service/saucelabs-daemon.ts [hunk 5]
```diff
@@ -195,20 +194,24 @@ export class SaucelabsDaemon {
...
    * @internal
-   * Launches all browsers. If there are pending tests waiting for a particular browser to launch
-   * before they can start, those tests are started once the browser is launched.
+   * Launches a set of browsers and increments the count of parallel browser started. If there are
+   * pending tests waiting for a particular browser to launch before they can start, those tests are
+   * started once the browser is launched.
    **/
-  private async _launchBrowsers() {
-    console.debug('Launching browsers...');
+  private async launchBrowsers() {
```
Comment (devversion) [inline tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:205]: just to make it clear that this is not launching _all_ possible browsers

```suggestion
  private async launchBrowserSet() {
```

File: tools/saucelabs-daemon/background-service/saucelabs-daemon.ts [hunk 6]
```diff
@@ -294,16 +299,31 @@ export class SaucelabsDaemon {
...
+
+        // If the browser is launching, check if it can be pre-claimed so that
+        // the test starts once the browser is ready. If it's already claimed,
+        // continue searching.
+        if (browser.state === 'launching' && this._pendingTests.has(browser)) {
+          continue;
+        }
+
+        match = browser;
+        break;
```
Comment (devversion) [inline tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:323]: Nit: Can return `match` directly I think?

File: tools/saucelabs-daemon/background-service/saucelabs-daemon.ts [hunk 7]
```diff
@@ -294,16 +299,31 @@ export class SaucelabsDaemon {
...
    **/
-  private _findMatchingBrowsers(browserId: string): RemoteBrowser[] {
-    const browsers: RemoteBrowser[] = [];
-    this._activeBrowsers.forEach(b => {
-      if (b.id === browserId) {
-        browsers.push(b);
+  private _findAvailableBrowser(browserId: string): RemoteBrowser|undefined {
+    let match: RemoteBrowser|undefined = undefined;
+    for (const browser of this._activeBrowsers) {
+      if (browser.id === browserId) {
```
Comment (devversion) [inline tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:309]: Nit: Can avoid the nesting by just continuing if `browser.id !== browserId`

**Review by devversion — APPROVED** [review]
LGTM overall. Thanks. Just to be clear we are still not on the minimum of browsers because we are always launching a "set" of parallelize-able browsers. e.g. might be that only Chrome is requested, but we are still launching Firefox too. In practice not worth optimizing for though- this is sufficient- because both Firefox&Chrome would run most of the time together anyway.. (so maybe even an optimization then :P)

File: tools/saucelabs-daemon/background-service/saucelabs-daemon.ts (hunk 2 above)
Comment (gregmagolan) [reply tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:306]: 👍

File: tools/saucelabs-daemon/background-service/saucelabs-daemon.ts (hunk 3 above)
Comment (gregmagolan) [reply tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:80]: Hmmm. Yes, `Browser[]` should work. That ens up just being the `Object.values` of CustomLaunchers.

File: tools/saucelabs-daemon/background-service/saucelabs-daemon.ts (hunk 4 above)
Comment (gregmagolan) [reply tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:174]: Sure. This should never happen so a log sounds good to me.

File: tools/saucelabs-daemon/background-service/saucelabs-daemon.ts (hunk 5 above)
Comment (gregmagolan) [reply tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:205]: 👍

File: tools/saucelabs-daemon/background-service/saucelabs-daemon.ts (hunk 6 above)
Comment (gregmagolan) [reply tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:323]: 👍

File: tools/saucelabs-daemon/background-service/saucelabs-daemon.ts (hunk 7 above)
Comment (gregmagolan) [reply tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:309]: 👍

File: .circleci/config.yml (hunk 1 above)
Comment (gregmagolan) [reply .circleci/config.yml]: removed

Comment (gregmagolan) [reply .circleci/config.yml]: last run before removing: https://app.circleci.com/pipelines/github/angular/angular/61319/workflows/c2c16c68-5eaf-4d86-8ffe-3b827359ebcc/jobs/1339104

**Review by devversion — APPROVED** [review]

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit 34d73019f3cf8eee23ff23bf5896034c8e54a675.

**Pull request closed without merging**